### ISSUE TITLE
Weird behavior in reduce function

### ISSUE THREAD

**Issue by ThibautGery** [ramda/ramda#2239]
I am experiencing different behavior with the `reduce` function.
Those two functions have different behaviors:
```js
const fromListToMap = R.reduce((acc, attr) => {
  acc[attr.name] = attr.value
  return acc
}, {})

const fromListToMapWithoutCurrying = list => R.reduce((acc, attr) => {
  acc[attr.name] = attr.value
  return acc
}, {})(list)
```
Input:
```json
[
  [
    { "name": "toto", "value": "toto1"}
  ],
  [
    { "name": "tata", "value": "tata1"}
  ],
]
```
Calling
```js
 R.map(fromListToMap)(input)
```
returns
```json
[
  {
    "toto": "toto1",
    "tata": "tata1"
  },
  {
    "toto": "toto1",
    "tata": "tata1"
  }
```
instead of 
```json
[
  {
    "toto": "toto1",
  },
  {
    "tata": "tata1"
  }
```
Is this a bug or an is it the expected behavior?

**Comment by kedashoe** [comment]
Hi @ThibautGery , this is happening because in the first example you create the object in the outer scope, so each call to `fromListToMap` is mutating the same object, whereas in the second you create a new object each time.

```js
let x = R.map(fromListToMap)(input);
x[0].foo = 10;
console.log(x); // => [
  {
    "toto": "toto1",
    "tata": "tata1",
    "foo": 10,
  },
  {
    "toto": "toto1",
    "tata": "tata1",
    "foo": 10,
  }
]
```
fwiw, these sorts of questions are fine to ask on github, but even better is our gitter room which is quite active https://gitter.im/ramda/ramda.

**Comment by ThibautGery** [comment]
From what I understand, with the function `fromListToMap `, the second argument is the same for each function call. Is it correct?
In my opinion, this is not really a pure function. It introduces some side effects.

_I will make sure to use gitter next time._

**Comment by kedashoe** [comment]
> the second argument is the same for each function call. Is it correct?

I would not say that it is the same, the second argument is the initial value, and it is re-assigned at every step of the reduction: https://github.com/ramda/ramda/blob/v0.24.1/src/internal/_reduce.js#L11.

This argument is passed to our reducing function. In your case, you are mutating that argument:

```js
acc[attr.name] = attr.value
```
, which is why you are seeing the behaviour that you are. If you were to rewrite it without using mutation, eg with `R.assoc`, which does not mutate its input:
```js
const fromListToMap = R.reduce((acc, attr) => {
  return R.assoc(attr.name, attr.value, acc);
}, {})

const fromListToMapWithoutCurrying = list => R.reduce((acc, attr) => {
  return R.assoc(attr.name, attr.value, acc);
}, {})(list)
```
then I think you would get what you initially expected: https://goo.gl/KvxVtP

**Comment by ThibautGery** [comment]
Thank you @kedashoe for taking the time to explain that to me.
Your examples are enlighting. I was able to reproduce the same behavior using an `Object.assign` which mimics an immutable object assignment like `R.assoc`

**The next part is only to open the discussion on this behavior that I find not straightforward.**

**TL;DR**: Ramda should help as much as possible to contains side effect to local scope.

According to me, functional programming is writing pure functions without any side effects. And because there is no side effect, it is easy to test, easy to compose, easy to play with.

In my code, the issue is the line of mutable code. The issue is not present in fully functional languages because the language doesn't accept any mutable operations like Haskell.

```js
const fromListToMap = R.reduce((acc, attr) => {
  acc[attr.name] = attr.value // mutable code
  return acc
}, {})
```

But the issue is that javascript allows it. You can not force the developer to write immutable code.
In addition, writing mutable code is not a necessary a problem, if your mutations are kept within a local scope, this is not a problem. The best example of that is the  [`List#takeWhile`](https://github.com/scala/scala/blob/419a6394045a0615cb996152b04c92d25f9fb700/src/library/scala/collection/immutable/List.scala#L356-L364) implementation in Scala. There are mutations but this are kept locally.
My issue, here, was: I don't know how `ramda#reduce` function was working internally. I thought that the accumulator was duplicated for each call. If I am to continue to work with Ramda I need to learn its internal tidbits before being really efficient with it.

**My point is**: Ramda should help as much as possible to contains side effect by duplicating every constant like the reduce accumulator because javascript allows mutation unlike Haskell.

To do so I reimplemented the `reduce` function the way I expected it to behave.

```js
const R2 = {
  reduce: (lambda, acc) => (array) => {
    let newAcc = JSON.parse(JSON.stringify(acc)) // ugly but does the job
    for (let value of array) {
      newAcc = lambda(newAcc, value)
    }
    return newAcc
  }
}
```
Full demo in [REPL](https://goo.gl/ixTwvT)

This is not production ready. this is just to demonstrate my point of view. I understand that you have more constraints to work with.

**So my question is**: Are you willing to accept any PR on the subject of limiting side effect with that kind of functions? Are there any technical limitations?

Last but not least, there is a lot of people jumping on the functional hype train (me included), would it be possible to add a section in the documentation with _do_ and _dont_ examples?

If you disagree with my writing, feel free to close this issue.

Cheers

**Comment by buzzdecafe** [comment]
I appreciate what you are trying to accomplish, and I support your desire to see these functions behave in a truly immutable way. However, I think bolting immutability onto a language that does not provide support for it -- in fact, the base language is actively *hostile* to it -- is a _very_ difficult challenge. 

Ultimately, there is no way to ensure that the reducing function is pure. And the stringify/parse solution for cloning is too naive. Suppose I am folding onto a FL-compliant Functor with `map` defined on the prototype? This is a common use case for Ramda users. That would fail *immediately*, since the clone would ignore prototype properties; and JSON does not serialize functions. Perf would already be bad using stringify/parse; now you have to accommodate cloning (or somehow reattaching) prototype properties as well?

So should you try to prove this concept, expect some skepticism. A comprehensive solution may not be so easy to reach.

**Comment by ThibautGery** [comment]
@buzzdecafe Thank you for the quick feedback.
You are right, the implementation `JSON.parse(JSON.stringify(acc))` is very naive but this was only for the purpose of the demonstration.

>  However, I think bolting immutability onto a language that does not provide support for it -- in fact, the base language is actively hostile to it -- is a very difficult challenge.

I definitely agree on that, this is why I had my weird behavior in the first place

>  Suppose I am folding onto a FL-compliant Functor with map defined on the prototype? This is a common use case for Ramda users. That would fail immediately, since the clone would ignore prototype properties; and JSON does not serialize functions.

I am not sure I understand correctly, is there a unit test covering the use case in the [test suite](https://github.com/ramda/ramda/blob/master/test/reduce.js)?
I will see what I can do and try to submit a PR on this subject. We can continue this discussion on a not so naive implementation.

**Comment by CrossEye** [comment]
Part of the issue is that Ramda is rarely in the business of enforcing anything.  It tries to make possible a functional paradigm, but it doesn't insist on it.  It tries to work as well as possible with normal JS practices.

This make it quite difficult to do what you're looking for.  Trying to make a `clone` function that is robust enough for OO use-cases is extremely difficult.  We punt on it in [`R.clone`][cl].  But to do what you want, you would need a powerful cloning method.

And there's another objection.  You want Ramda to protect you from mutating your accumulator.  If we did, how would we leave it open for that (hopefully rare) user who specifically **wants** to update an accumulator object in each iteration?

```js
reduce(
  (state, item) => {state[item.name] = item.value; return state},
  globalStateObject, // I have no control over this; it's a reference passed by an outside 
                     // system, but I still need to update it.
  [{name: 'foo', value: 'bar'}, {name: 'baz', value: 'qux'}]
)
```

We could of course say, "Don't do that."  But I think then we'd be writing a different library, or perhaps a different language.


-----
> I am not sure I understand correctly, is there a unit test covering the use case in the [test suite][ts]?

Yes, [this one][te]:

```js
  it('dispatches to objects that implement `reduce`', function() {
    var obj = {x: [1, 2, 3], reduce: function() { return 'override'; }};
    eq(R.reduce(add, 0, obj), 'override');
    eq(R.reduce(add, 10, obj), 'override');
  });
```

----------

> **So my question is**: Are you willing to accept any PR on the subject of limiting side effect with that kind of functions? Are there any technical limitations?

I think the restrictions you'll find are along the lines @buzzdecafe suggested: it's a _very_ difficult problem, and the core team might not be convinced it's worth doing even if you could manage it.


> Last but not least, there is a lot of people jumping on the functional hype train (me included), would it be possible to add a section in the documentation with _do_ and _dont_ examples?

For that, a PR would be **most** welcome!

  [cl]: http://ramdajs.com/docs/#clone
  [te]: https://github.com/ramda/ramda/blob/103543e2e1c2b8a43275fff99c9e62363d1def85/test/reduce.js#L13-L17
  [ts]: https://github.com/ramda/ramda/blob/master/test/reduce.js
